sdks/csharp/adtIotHub_allocate.cs

```diff
@@ -106,38 +106,40 @@ public static async Task<string> FindOrCreateTwinAsync(string dtmi, string regId
                 });
 
             // Find existing twin with registration ID
-            string query = $"SELECT * FROM DigitalTwins T WHERE $dtId = '{regId}' AND IS_OF_MODEL('{dtmi}')";
-            AsyncPageable<BasicDigitalTwin> twins = client.QueryAsync<BasicDigitalTwin>(query);
-            string dtId = "";
-
-            await foreach (BasicDigitalTwin digitalTwin in twins)
-            {
-                dtId = digitalTwin.Id;
-                log.LogInformation($"Twin '{dtId}' with Registration ID '{regId}' found in DT");
-                break;
-            }
-
-            if (String.IsNullOrWhiteSpace(dtId))
+            
+            try
             {
-                // Not found, so create new twin
-                log.LogInformation($"Twin ID not found - setting DT ID to regID");
-                dtId = regId; // use the Registration ID as the DT ID
+                // Get twin with Id `regId`
+                BasicDigitalTwin dt = await GetDigitalTwin<BasicDigitalTwin>(regId).ConfigureAwait(false);
 
-                // Initialize the twin properties
-                var digitalTwin = new BasicDigitalTwin
+                // Check to make sure it is of model type `dtmi`
+                if (dt.Metadata.ModelId.Equals(dtmi, StringComparison.OrdinalIgnoreCase)
                 {
-                    Metadata = { ModelId = dtmi },
-                    Contents =
+                    return dt.Id;
+                }
+                else
```

Comment:
The question is, do we want to override the existing digital twin if the model is different.

Reply:
currently we return null with this PR.

Reply:
The calling code doesn't handle null. Either it needs to, or we should revert to the original behavior.

Reply:
I think it would be interesting to add another example (even if in comments) that says, another way of doing this is to use `CreateOrReplaceDigitalTwinAsync` using `ETag.All`, which effectively overwrites the existing item if it is there, and creates it if it isn't.

That assumes you don't conditionally created it based on the dtmi, though.

Reply:
I will revert it to the original behavior and replace the twin, and make it clear that we are doing so.

Reply:
With the latest commit as of yet, here is the behavior that matches the behavior previously

if we find a digital twin with the same id and of the same model `dtmi` we will return the id of that dt.
if we find a digital twin with the same id and not of the same model `dtmi` we will replace the existing one with a new digitaltwin of the model `dtmi`
if we don't find a digital twin with the specified id, we will create a new one and return the id.

Reply:
Agree on this behavior, as it will update the twin to the correct state. The outcome of all three conditions will then be as expected.